clang/lib/Headers/__clang_cuda_math_forward_declares.h

```diff
@@ -70,6 +70,9 @@ __DEVICE__ double floor(double);
 __DEVICE__ float floor(float);
 __DEVICE__ double fma(double, double, double);
 __DEVICE__ float fma(float, float, float);
+#ifdef _MSC_VER
+__DEVICE__ long double fma(long double, long double, long double);
```

Comment:
That's right, CUDA has provided `long double fma` since 10.2 but for `HOST` only:

```cpp
inline _LIBCUDACXX_INLINE_VISIBILITY long double fma(long double __lcpp_x, long double __lcpp_y, long double __lcpp_z) _NOEXCEPT {return ::fmal(__lcpp_x, __lcpp_y, __lcpp_z);}
```

If a declaration of `device long double fma` is needed in order to parse it only and with disallowing its usage on GPU, then it is better to add it to `clang/lib/Headers/cuda_wrappers/cmath` under `_LIBCPP_STD_VER` and `_LIBCPP_HIDE_FROM_ABI`; and it would need a corresponding `GPU` test, of course.

Reply:
We already have a handful of MSVC-specific no-implementation functions declared here, so one more is OK.
I just want to document it better.

You may want to add a macro with a descriptive name. (E.g. `CUDA_ALLOW_LONG_DOUBLE_MATH_FUNCTION_DECLS`) and have it defined for MSVC, along with the comment why it's needed. Then replace the existing `#if _MSC_VER` with it.

Reply:
Yes, we have. But how do we prevent them from running on GPU?

Reply:
Given that there's no implementation for these functions, the reference will remain unresolved and GPU-side executable linking will fail.

Reply:
That is it, it is only a compile-time story, and the situation when the successfully compiled code "fails somewhere in ptxas" on runtime would never happen till CUDA implementation of `long double fma` appears, or am I missing something?

Reply:
Correct. If someone ends up using fma(long double) in the GPU-side code, we will fail at build time, until the implementation is actually available. Failure in ptxas is not the best way to diagnose it, but it's better than failing to compile valid code because of a CUDA compatibility quirk in a standard library.